### ISSUE TITLE
Inconsistent gradients in reverse mode and forward mode AD when using `MultiheadAttention`

### ISSUE THREAD

**Issue by LouisChenB** [pytorch/pytorch#126141]
### 🐛 Describe the bug

When I use forward and reverse AD(automatic differentiation), the calculated gradients are not consistent. I'm wondering what's goning on.
```
from helper_torch import (
    RevInv,
    FwdInv,
)
from classes.torch_library import TorchLibrary
import torch


torch.random.manual_seed(44298)
def get_fn():
    embed_dim = 512
    num_heads = 16
    dropout = True
    arg_class = torch.nn.MultiheadAttention(embed_dim, num_heads, dropout=dropout, device='cpu', )
    def fn(input_0, input_1, input_2):
        fn_res = arg_class(input_0, input_1, input_2)
        return fn_res
    return fn
fn = get_fn()
input_0_tensor = torch.empty([1, 1, 512], dtype=torch.float32)
input_0_tensor.uniform_(0, 7)
input_1_tensor = torch.empty([1, 1, 512], dtype=torch.float32)
input_1_tensor.uniform_(-2, 1)
input_2_tensor = torch.empty([1, 1, 512], dtype=torch.float32)
input_2_tensor.uniform_(-1, 7)


inputs = (input_0_tensor,input_1_tensor,input_2_tensor)

rev_status, rev_value, rev_grad, rev_err = RevInv(fn, inputs)
fwd_status, fwd_value, fwd_grad, fwd_err = FwdInv(fn, inputs)
print(TorchLibrary.is_equal(rev_grad, fwd_grad))

total_difference = 0
for grad1, grad2 in zip(rev_grad, fwd_grad):
    if grad1 is not None and grad2 is not None:
        if grad1.shape == grad2.shape:
            difference = torch.norm(grad1 - grad2)
            total_difference += difference.item()
        else:
            print(grad1.shape, grad2.shape)
print("total_difference: ", total_difference)
# False
# torch.Size([512, 512]) torch.Size([512, 1])
# torch.Size([512, 1]) torch.Size([512, 512])
# total_difference:  0.0
```

### Versions

PyTorch: 2.2.0
Ubuntu: 20.04
Python: 3.10.12

**Comment by albanD** [comment]
Can you share what you RevInv and FwdInv functions are?

**Comment by LouisChenB** [comment]
```
def RevInv(fn, inputs, device="cpu"):
    value = None
    err_msg = ""
    gradient = None
    try:
        inputs_0 = CopyInputs(inputs, grad=True, device=device)
        value = fn(*inputs_0)
        backward(value)
        inputs_1 = CopyInputs(inputs, grad=True, device=device)
        gradient = get_jacobians(fn, inputs_1, backward_ad=True)
    except Exception as e:
        status = "fail"
        err_msg = str(e)
    else:
        status = "success"
    return status, value, gradient, err_msg


def FwdInv(fn, inputs, device="cpu"):
    value = None
    err_msg = ""
    gradient = None
    try:
        inputs_0 = CopyInputs(inputs, grad=True, device=device)
        with fwad.dual_level():
            dual_inputs = []
            for inp in inputs_0:
                tan = torch.rand_like(inp)
                dual_inputs += [fwad.make_dual(inp, tan)]
            out = fn(*dual_inputs)
            if isinstance(out, tuple):
                value = []
                for o in out:
                    temp, _ = fwad.unpack_dual(o)
                    value.append(o)
                value = tuple(value)
            else:
                value, _ = fwad.unpack_dual(out)
        inputs_1 = CopyInputs(inputs, grad=True, device=device)
        gradient = get_jacobians(fn, inputs_1, backward_ad=False)
    except Exception as e:
        status = "fail"
        err_msg = str(e)
    else:
        status = "success"
    return status, value, gradient, err_msg
```

**Comment by LouisChenB** [comment]
You can use the command "python rev_fwd.py" at this link
https://hz-t3.matpool.com:26998?token=aebKc5hhry

**Comment by albanD** [comment]
Your foward ad tangents are random: `tan = torch.rand_like(inp)`. How do you ensure the backward ad is doing the same projections?

Also can you share you code as gist to make sure this stays and not downloading from untrusted source?
